org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/processing/JobManager.java

```diff
@@ -29,32 +31,54 @@
 
 public abstract class JobManager implements Runnable {
 
-	/* queue of jobs to execute */
-	protected List<IJob> awaitingJobs = new LinkedList<>();
+	/**
+	 * queue of jobs to execute
+	 * <br>
+	 * synchronized by JobManager.this
+	 */
+	private List<IJob> awaitingJobs = new LinkedList<>();
+
+	private volatile boolean executing;
 
-	protected volatile boolean executing;
+	/**
+	 * background processing
+	 * <br>
+	 * synchronized by JobManager.this
+	 */
+	private Thread processingThread;
 
-	/* background processing */
-	protected volatile Thread processingThread;
-	protected volatile Job progressJob;
+	private volatile Job progressJob;
 
-	/* counter indicating whether job execution is enabled or not, disabled if <= 0
-	    it cannot go beyond 1 */
+	/**
+	 * counter indicating whether job execution is enabled or not, disabled if <= 0
+	 * it cannot go beyond 1
+	 * <br>
+	 * synchronized by JobManager.this
+	 */
 	private int enableCount = 1;
 
 	public static boolean VERBOSE = false;
-	/* flag indicating that the activation has completed */
-	public boolean activated = false;
 
-	private int awaitingClients = 0;
+	/**
+	 * flag indicating that the activation has completed
+	 * <br>
+	 * synchronized by JobManager.this
+	 */
+	private boolean activated;
+
+	private AtomicInteger awaitingClients = new AtomicInteger();
 
 	private final Object idleMonitor = new Object();
 
+	private synchronized Thread getProcessingThread() {
+		return this.processingThread;
+	}
 	/**
 	 * Invoked exactly once, in background, before starting processing any job
 	 */
-	public void activateProcessing() {
+	public synchronized void activateProcessing() {
 		this.activated = true;
+		this.notifyAll(); // someone may wait for awaitingJobsCount()
```

Comment:
Not sure this call is needed here (and synchronized method call), as it doesn't necessarily means there job count is changed - it **may** change later on.

Reply:
awaitingJobsCount reads activated which is changed here